oak/server/rust/oak_glue/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use byteorder::{ReadBytesExt, WriteBytesExt};
+use lazy_static::lazy_static;
+use log::{debug, error, info, warn};
+use oak_abi::OakStatus;
+use oak_runtime::proto::ApplicationConfiguration;
+use oak_runtime::runtime::RuntimeProxy;
+use oak_runtime::NodeId;
+use prost::Message;
+use std::convert::TryInto;
+use std::io::Cursor;
+use std::sync::{Arc, RwLock};
+
+#[no_mangle]
+pub extern "C" fn glue_init() {
+    let _ = ::std::panic::catch_unwind(|| {
+        simple_logger::init().expect("failed to initialize logger");
+        info!("Rust FFI glue initialized");
+    });
+}
+
+type NodeFactory =
+    extern "C" fn(data: usize, name: *const u8, name_len: u32, node_id: u64, handle: u64) -> ();
+
+struct Glue {
+    runtime: Arc<oak_runtime::Runtime>,
+    factory: Option<NodeFactory>,
+    factory_data: usize,
+}
+
+impl Glue {
+    fn new(
+        runtime: Arc<oak_runtime::Runtime>,
+        factory: Option<NodeFactory>,
+        factory_data: usize,
+    ) -> Self {
+        Glue {
+            runtime,
+            factory,
+            factory_data,
+        }
+    }
+    /// Recreate a RuntimeProxy instance that corresponds to the given node ID
+    /// value.
+    fn proxy(&self, node_id: u64) -> RuntimeProxy {
+        RuntimeProxy {
+            runtime: self.runtime.clone(),
+            node_id: NodeId(node_id),
+        }
+    }
+}
+
+lazy_static! {
+    static ref GLUE: RwLock<Option<Glue>> = RwLock::new(None);
+}
+
+const R1: &str = "global glue lock poisoned";
+const R2: &str = "global glue object missing";
+
+/// Start the Rust runtime, with the ApplicationConfiguration provided in
+/// serialized form.
+///
+/// # Safety
+///
+/// Caller must ensure that the memory range [config_buf, config_buf+config_len) is
+/// accessible and holds a protobuf-serialized ApplicationConfiguration message.
+#[no_mangle]
+pub unsafe extern "C" fn glue_start(
```

Comment:
Could `unsafe` be moved inside the body only in the sections where it is actually necessary?

Reply:
Yes, the function is explicitly marked unsafe, as this "warns its callers that the function has a contract that they must satisfy to avoid undefined behavior" (Programming Rust, p.531).  (Aside: marking it `unsafe` triggers a lint rule that the doc comment has to have a "Safety" section).

Reply:
Yeah, that's what I would expect.